src/index.ts

```diff
@@ -82,17 +90,23 @@ export class EIP7412 {
             oracleQuery
           );
 
-          multicallCalls.splice(multicallCalls.length - 1, 0, {
-            to: err.args![0] as viem.Address,
-            data: viem.encodeFunctionData({
-              abi: IERC7412.abi,
-              functionName: "fulfillOracleQuery",
-              args: [signedRequiredData],
-            }),
-          });
+          multicallCalls.splice(
+            multicallCalls.length - initialMulticallLength,
+            0,
+            {
+              to: err.args![0] as viem.Address,
+              data: viem.encodeFunctionData({
+                abi: IERC7412.abi,
+                functionName: "fulfillOracleQuery",
+                args: [signedRequiredData],
+              }),
+            }
+          );
         } else if (err.errorName === "FeeRequired") {
           const requiredFee = err.args![0] as bigint;
-          multicallCalls[multicallCalls.length - 2].value = requiredFee;
+          multicallCalls[
+            multicallCalls.length - initialMulticallLength - 1
+          ].value = requiredFee;
```

Comment:
Say your tx requires two price updates.
- Do tx
- OracleDataRequired
- Add price update second last
- Do muti tx
- OracleDataRequired again
- Add price update second last
- FeeRequired (from the first one), overwrite value on second to last
- Do multi tx
- FeeRequired (from the second one), overwrite value on second to last. Here I think we have a problem, we will overwrite the fee required from the previous one. I'd expect to be in a FeeRequired infinity loop

Reply:
not sure what you mean by "second to last". Multicalls are always prepended to the initial list of calls given. Each call in the multicall sends its own "value" to the contract requesting price update (and the toal amount for the required is simply the sum of all the `value`s for all the invidiual txns on the multicall). so based on the math that I am seeing here

btw we have a test verifying the case you are talking about `test/client.mjs` on the repo. it verifies multiple oracledataRequireds being returned, all with fee required, so if the math was not right heren I would expect that test would stop working.

Reply:
That sounds reassuring. I'll have a another look

Reply:
I'm a bit under the gun today, but will look into it more when I have time.. When I'm reading the logic it still fees wrong, but I could def be misunderstanding `FeeRequired` error. This is unrelated to this change and would've been an issue before too. This would only be an issue if the original tx required two price updates.

The way I'm reading it.
Start out with a single tx:
-> `multicallCalls = [initialTx]`
OracleDataRequired: preprend price update
-> `multicallCalls = [priceUpdate, initialTx]`
OracleDataRequired: preprend price update 2 
-> `multicallCalls = [priceUpdate, priceUpdate, initialTx]`
Fee Required (first price update)
-> Update `value` for tx at index: `multicallCalls.length - initialMulticallLength` = 3 - 1 = 2
Fee Required (second price update)
-> Update `value` for tx at index: `multicallCalls.length - initialMulticallLength` = 3 - 1 = 2`
This will update the same tx?
```
const txToUpdate =multicallCalls[multicallCalls.length - initialMulticallLength - 1]
txToUpdate.value = txToUpdate.value + requiredFee
```
In my mind this would fix the issue..

I guess if the first FeeRequired have enough value for both price update we will never get the second Fee Required and we dont have a problem and then I thank you for being patient with me...

Reply:
the mistake in your example occurs because the `prepend price update 2` occurs after `Fee Required (first price update)`. This is because, if you think about it, by the time the second price update is being set up, the first price update has been fully processed its function through (including any fees it may require)

Reply:
Cool, that does make sense!